src/Microsoft.AspNet.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs

```diff
@@ -24,61 +25,88 @@ public CompositeViewEngine(IOptions<MvcViewOptions> optionsAccessor)
         public IReadOnlyList<IViewEngine> ViewEngines { get; }
 
         /// <inheritdoc />
-        public ViewEngineResult FindPartialView(
-            ActionContext context,
-            string partialViewName)
+        public ViewEngineResult FindView(ActionContext context, string viewName, bool isMainPage)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            if (partialViewName == null)
+            if (string.IsNullOrEmpty(viewName))
             {
-                throw new ArgumentNullException(nameof(partialViewName));
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(viewName));
             }
 
-            return FindView(context, partialViewName, partial: true);
-        }
-
-        /// <inheritdoc />
-        public ViewEngineResult FindView(
-            ActionContext context,
-            string viewName)
-        {
-            if (context == null)
+            // Do not allocate in the common cases: ViewEngines contains one entry or initial attempt is successful.
+            IEnumerable<string> searchedLocations = null;
+            List<string> searchedList = null;
+            for (var index = 0; index < ViewEngines.Count; index++)
```

Comment:
1. Because this is two lines after a comment saying what follows is allocation free. But `foreach` allocates at least the enumerator.
2. I like `index` :smiley_cat: but will switch if you feel strongly. Do you?